lib/common/xml_display.c

```diff
@@ -206,30 +200,55 @@ log_xml_node(GString *buffer, int log_level, const char *prefix,
 
 /*!
  * \internal
- * \brief Log an XML element or comment in a formatted way
+ * \brief Output an XML element or comment in a formatted way
  *
- * Depending on the value of \p log_level, the output may be written to
- * \p stdout or to a log file.
+ * \param[in,out] out        Output object
+ * \param[in]     log_level  Priority at which to log the messages (ignored if
+ *                           \p out is not \p NULL)
+ * \param[in]     prefix     String to prepend to every line of output
+ * \param[in]     data       XML node to output
+ * \param[in]     depth      Current nesting level
+ * \param[in]     options    Group of \p pcmk__xml_fmt_options flags
  *
- * \param[in] log_level  Priority at which to log the messages
- * \param[in] prefix     String to prepend to every line of output
- * \param[in] data       XML node to log
- * \param[in] depth      Current indentation level
- * \param[in] options    Group of \p pcmk__xml_fmt_options flags
+ * \note This currently produces output only for text-like output objects.
  */
 void
-pcmk__xml_log(int log_level, const char *prefix, const xmlNode *data, int depth,
-              uint32_t options)
+pcmk__xml_show(pcmk__output_t *out, int log_level, const char *prefix,
+               const xmlNode *data, int depth, uint32_t options)
 {
-    /* Allocate a buffer once, for log_xml_node() to truncate and reuse in
+    GString *buffer = NULL;
+    bool need_output_free = false;
+
+    if (out == NULL) {
+        switch (log_level) {
+            case LOG_NEVER:
+                return;
+            case LOG_STDOUT:
+                CRM_CHECK(pcmk__text_output_new(&out, NULL) == pcmk_rc_ok,
+                          return);
+                break;
+            default:
+                CRM_CHECK(pcmk__log_output_new(&out) == pcmk_rc_ok, return);
+                pcmk__output_set_log_level(out, log_level);
+                break;
+        }
+
+        need_output_free = true;
+    }
+
+    CRM_CHECK(depth >= 0, depth = 0);
+
+    /* Allocate a buffer once, for output_xml_node() to truncate and reuse in
      * recursive calls
      */
-    GString *buffer = g_string_sized_new(1024);
+    buffer = g_string_sized_new(1024);
 
-    CRM_CHECK(depth >= 0, depth = 0);
+    output_xml_node(out, buffer, prefix, data, depth, options);
 
-    log_xml_node(buffer, log_level, prefix, data, depth, options);
     g_string_free(buffer, TRUE);
+    if (need_output_free) {
+        pcmk__output_free(out);
```

Comment:
need to finish the output object before freeing it, if we created the output object